File: 362. Design Hit Counter.py
class HitCounter:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.q = [(0,0) for _ in range(300)]

    def hit(self, timestamp: int) -> None:
        """
        Record a hit.
        @param timestamp - The current timestamp (in seconds granularity).
        """
        index = timestamp % 300
        time, hit = self.q[index]
        if timestamp != time:
            self.q[index] = (timestamp, 1)
        else:
            self.q[index] = (time, hit+1)

    def getHits(self, timestamp: int) -> int:
        """
        Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity).
        """
        count = 0
        for t, hit in self.q:
            if t > timestamp-300:
                count += hit
        return count


# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)

File: 362. Design Hit Counter.java
class HitCounter {
    Deque<Integer> dq;

    /** Initialize your data structure here. */
    public HitCounter() {
        dq = new ArrayDeque<>();
    }
    
    /** Record a hit.
        @param timestamp - The current timestamp (in seconds granularity). */
    public void hit(int timestamp) {
        dq.offer(timestamp);
    }
    
    /** Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity). */
    public int getHits(int timestamp) {
        while (!dq.isEmpty() && dq.peekFirst() <= timestamp-300) {
            dq.pollFirst();
        }
        return dq.size();
    }
}

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter obj = new HitCounter();
 * obj.hit(timestamp);
 * int param_2 = obj.getHits(timestamp);
 */